### PR TITLE
Django1.11

### PR DESCRIPTION
**Pull request by mmanogaran**
Files in apps/api are now working. Everything was updated to Django 1.11

### CODE REVIEW

File: apps/api/admin.py [hunk 1]
```diff
@@ -1,85 +1,170 @@
...
 
 
 class IdentificationInline(admin.StackedInline):
+    '''Making it StackedInline means that this interface can also be
+    edited by other interfaces
+    '''
     model = models.Identification
 
 
+class IdentificationAdmin(admin.ModelAdmin):
```
Comment (mmanogaran) [inline apps/api/admin.py:39]: Many unnecessary admin classes that only passes.

Possible Solution: Remove this and only have admin.site.register(models.Identification) in this case. Modify accordingly for other models.

File: apps/api/models.py [hunk 2]
```diff
@@ -40,37 +48,61 @@ class Identification(models.Model):
...
+        return self.code
+
     def __unicode__(self):
         return self.code
 
 
 class Assignment(models.Model):
     name = models.CharField(max_length=254)
-    rubric_entries = models.ManyToManyField("Rubric", blank=True, related_name="rubric_entries")
+    rubric_entries = models.ManyToManyField(
```
Comment (mmanogaran) [inline apps/api/models.py:101]: Is a many-to-many field required for rubrics? An assignment has many rubrics. Does a rubric have many assignments?

File: apps/api/models.py [hunk 3]
```diff
@@ -79,7 +111,10 @@ def __unicode__(self):
 class Rubric(models.Model):
     name = models.CharField(max_length=254)
     total = models.DecimalField(max_digits=6, decimal_places=2)
-    assignment = models.ForeignKey("Assignment")
+    assignment = models.ForeignKey("Assignment", blank=True)
```
Comment (mmanogaran) [inline apps/api/models.py:114]: Following Assignment comment. There is an Assignment ForeignKey here. I don’t think it is necessary with the many-to-many field in Assignment.  This seems to be the global variable issue mentioned before.

File: apps/api/validators.py [hunk 4]
```diff
@@ -1,12 +1,384 @@
-from django.core.exceptions import ValidationError
+'''Validates if the file given is a csv file'''
 
 import csv
+from re import match
+from django.core.exceptions import ValidationError
+
+
+class ColumnError(Exception):
```
Comment (mmanogaran) [inline apps/api/validators.py:8]: Currently just inherits from Exception and passes. Every time it is raised, it uses the same message with the same parameters : current row number, expected column number, and row’s number of column. Will be troublesome to find and modify all if something changes.

Possible solution (using same variable names and message):

	class ColumnError(Exception):
		def __init__(self, line_number, col_number, len_row):
			msg: ‘Row: {}. Expected: {} columns. Found {} columns’.format(line_number, col_number, len_row)
			super(ColumnError, self).__init__(msg)

EX. raise ColumnError(8, 4, 2)
Row: 8. Expected: 4 columns. Found 2 columns

File: apps/api/validators.py [hunk 5]
```diff
@@ -1,12 +1,384 @@
...
+    '''Exception raised if the number of columns in the csv file header does
+    not match the rest'''
+    pass
+
+
+class MultipleUtoridOccurences(Exception):
```
Comment (mmanogaran) [inline apps/api/validators.py]: Same as ColumnError except passing in (‘, ‘.join(error_utorid))

Possible solution:

```
class MultipleUtoridOccurences(Exception):
	def __init__(self, error_utorid):
		msg = ‘, ‘.join(error_utorid)
		super(MultipleUtoridOccurences, self).__init__(msg)
```

raise MultipleUtoridOccurences(error_utorid)

File: apps/api/validators.py [hunk 6]
```diff
@@ -1,12 +1,384 @@
...
+    '''Exception raised if the same utor id occurs multiple times in a file
+    '''
+    pass
+
+
+class InvalidUtorid(Exception):
```
Comment (mmanogaran) [inline apps/api/validators.py]: Same as ColumnError. Always passing in current row number and same message.

Possible solution:

```
class InvalidUtorid(Exception):
	def __init__(self, line_number):
		msg = ‘Invalid utorid at row: {}, column: 0’.format(line_number)
		super(InvalidUtorid, self).__init__(msg)
```

Ex. raise InvalidUtorid(4)
Invalid utorid at row: 4, column: 0’

File: apps/api/validators.py [hunk 7]
```diff
@@ -1,12 +1,384 @@
...
+    '''Utor id not valid. It has to be 8 characters long and only contain
+    alphanumberic symbols
+    '''
+    pass
+
+
+class InvalidData(Exception):
```
Comment (mmanogaran) [inline apps/api/validators.py]: Similar to ColumnError. Varying data name and row number. Sometimes column number is passed in.

Possible Solution:

```
class InvalidData(Exception):
	def __init__(self, data_name, row, col=None):
		msg = ‘Invalid {} at row: {}’.format(data_name, row)
		if col:
			msg += ‘, column: {}.format(col)
		super(InvalidData, self).__init__(msg)
```
Ex. raise InvalidData(“last name”, 4)
Invalid last name at row: 4

Ex. raise InvalidData(“utorid”, 8, 3)
Invalid utorid at row: 8, column: 3

File: apps/api/validators.py [hunk 8]
```diff
@@ -1,12 +1,384 @@
...
+    '''If the data at a given column does not have the proper structure or
+    seem valid
+    '''
+    pass
+
+
+def _is_lowercase_alphabet(value):
```
Comment (mmanogaran) [inline apps/api/validators.py]: match() returns None is false and cause function to return AttributeError: ‘NoneType’ object has no attribute ‘group’

Possible solution:

```
	_is_lowercase_alphabet(value):
		return (value.isalpha() and value.islower())
```
Ex. _is_lowercase_alphabet(“”)
False

File: apps/api/validators.py [hunk 9]
```diff
@@ -1,12 +1,384 @@
...
+        value: str
+
+    Returns:
+        A boolean that determines if the parameter 'value' is completely
+        lowercase or not
+    '''
+    return len(match('[a-z]+', value).group(0)) == len(value)
+
+
+def _is_alphabet(value):
```
Comment (mmanogaran) [inline apps/api/validators.py]: Same as _is_lowercase_alphabet.

Possible solution:

```
	_is_alphabet(value):
		return value.isalpha()
```
Ex. _is_alphabet(“!”)
False

File: apps/api/validators.py [hunk 10]
```diff
@@ -1,12 +1,384 @@
...
+        value: str
+
+    Returns:
+        A boolean that determines if the parameter 'value' is composed
+        entirely of alphabet characters
+    '''
+    return len(match('[A-Za-z]+', value).group(0)) == len(value)
+
+
+def _is_alphanumeric(value):
```
Comment (mmanogaran) [inline apps/api/validators.py]: Same as _is_lowercase_alphabet.

Possible solution:
```
	_is_alphanumeric(value):
		return value.isalnum()
```

File: apps/api/validators.py [hunk 11]
```diff
@@ -1,12 +1,384 @@
...
+        value: str
+
+    Returns:
+        A boolean that determines if the parameter 'value' is composed of
+        alphabet and numeric characters
+    '''
+    return len(match('[A-Za-z0-9]+', value).group(0)) == len(value)
+
+
+def _is_alphanumeric_and_symbols(value):
```
Comment (mmanogaran) [inline apps/api/validators.py]: Same as _is_lowercase_alphabet.

Possible solution:

```
	_is_alphanumeric_and_symbols(value):
		match = match(‘[A-Za-z0-9_.!@#$%^&*()\’ ]+’, value)
		if not match:
			return False
		return len(match.group(0)) == len(value)
```

File: apps/api/validators.py [hunk 12]
```diff
@@ -1,12 +1,384 @@
...
+        value: str
+
+    Returns:
+        A boolean that determines if the parameter 'value' is composed of
+        alphabet, numeric and certain symbol characters
+    '''
+    return len(match('[A-Za-z0-9_.!@#$%^&*()\' ]+', value).group(0)) == len(value)
+
+
+def _is_numeric(value):
```
Comment (mmanogaran) [inline apps/api/validators.py]: Same as _is_lowercase_alphabet

Possible solution:

```
	_is_numeric(value):
		try:
			int(value)
		catch ValueError:
			return False
		return True
```

File: apps/api/validators.py [hunk 13]
```diff
@@ -1,12 +1,384 @@
...
+        value: str
+
+    Returns:
+        A boolean that determines if the parameter 'value' is a valid utorid.
+        Utor id is 8 characters in length and composed of alphabet and number
+    '''
+    return len(value) == 8 and _is_alphanumeric(value)
+
+
+def _valid_student_number(value):
```
Comment (mmanogaran) [inline apps/api/validators.py:153]: Returns true if value is alphanumeric.

Possible solution:

	Use numeric(value) instead of alphanumeric(value)

**Comment by bhowmikp** [comment]
Issues identified on the pull request were discussed and changes were made accordingly

**Pull request closed without merging**